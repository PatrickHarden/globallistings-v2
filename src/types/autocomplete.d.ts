import { Dictionary } from './common/dictionary';
import { Property } from './property';

// TODO remove the extra properties (except for isProperty) and move them into a property search interface to store in state
interface Autocomplete {
  readonly placeId: string;
  readonly description: string;
  readonly propertyResult?: PropertyResult;
}

interface PropertyResult {
  readonly propertyPartialPath?: string;
  readonly propertyAspects?: string[];
  readonly searchPath?: string;
}

interface NormalisedAutocompleteResult {
  readonly entities: AutocompleteEntities;
  readonly result: string[];
}

interface AutocompleteEntities {
  readonly suggestions: Dictionary<Autocomplete>;
}
