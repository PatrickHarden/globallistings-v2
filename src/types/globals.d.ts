declare module 'react-google-maps/lib/components/addons/MarkerWithLabel';
declare module 'lodash/debounce';

interface Window {
  dataLayer: any;
  devToolsExtension: any;
  __REDUX_DEVTOOLS_EXTENSION__: any;
  google: any;
}
