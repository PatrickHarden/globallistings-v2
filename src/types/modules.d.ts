import 'ramda';

declare module 'ramda' {
  interface Static {
    composeWith: (fn: (a: any, b: any) => any) => any;
  }
}
