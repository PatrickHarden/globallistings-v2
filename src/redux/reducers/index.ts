import entities from './entities';
import entitiesMeta from './entitiesMeta';
import ui from './ui';

export default {
  entities,
  entitiesMeta,
  ui,
};
