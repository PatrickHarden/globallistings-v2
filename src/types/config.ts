import { Dictionary } from './common/dictionary';

interface BaseConfig {
  readonly api: string;
  readonly bitlyToken: string;
  readonly breakpoints: Breakpoints;
  readonly cdnUrl: string;
  readonly filters: Filter[];
  readonly language: string;
  readonly largePlaceholderImageUrl: string;
  readonly mapMarkerIconUrl: string;
  readonly mapMarkerIconUrlInactive: string;
  readonly mediumPlaceholderImageUrl: string;
  readonly params: Params;
  readonly pdf: Pdf;
  readonly pdpContentDrivenBlock?: any;
  readonly pdpContentDrivenBlockLettings?: any;
  readonly placesAddressLookupStringFormat: string;
  readonly propertyContactApiUrl: string;
  readonly recaptchaKey: string;
  readonly searchConfig: SearchConfig;
  readonly searchLocationName?: string;
  readonly searchMode: string;
  readonly siteId: string;
  readonly siteType: string;
  readonly smallPlaceholderImageUrl: string;
  readonly tenantFeesLinkUrl?: string;
  readonly title: string;
  readonly urlPropertyAddressFormat: string;
  readonly currentConfig?: string;
  readonly limitListMapResults?: string;
  readonly otherConfigs?: OtherConfig[];
}

/**
 * These types should only be used by selectors and not by components directly
 * selectors need to output their own types
 */

export interface Config extends BaseConfig {
  readonly i18n: object;
}

export interface AppConfig extends BaseConfig {
  readonly i18n: Dictionary<string>;
}

export interface FilterOption {
  readonly value: string;
  readonly label: string;
  readonly default: boolean;
  readonly type?: string;
}

export enum FilterPlacement {
  group = 'lm_groupFilters',
  location = 'lm_locationFilter',
  primary = 'lm_primaryFilter',
  searchPrimary = 'primary',
  searchSecondary = 'secondary',
  secondary = 'lm_secondaryFilter',
  sort = 'lm_sortFilter',
  tertiary = 'lm_tertiaryFilters',
}

export enum FilterType {
  select = 'select',
  range = 'range',
}

export enum FilterModifier {
  query = 'query',
  url = 'url',
}

export type FilterCondition = { [K in keyof Params]?: string };

export interface BaseFilter {
  readonly placement: FilterPlacement;
  readonly modifier: FilterModifier;
  readonly name: string;
  readonly label: string;
  readonly conditional?: FilterCondition;
}

export interface SelectFilter extends BaseFilter {
  readonly options: FilterOption[];
  readonly type: FilterType.select;
}

export interface RangeFilter extends BaseFilter {
  readonly includePOA?: boolean;
  readonly maxValues: FilterOption[];
  readonly minValues: FilterOption[];
  readonly type: FilterType.range;
}

export type Filter = SelectFilter | RangeFilter;

interface Breakpoints {
  readonly xsmall: string;
  readonly small: string;
  readonly medium: string;
  readonly large: string;
}

interface Pdf {
  readonly staticMapApi: string;
  readonly pdfDownloadApi: string;
  readonly renderMiddleware: string;
}

export interface Params {
  readonly 'Common.Aspects': string;
  readonly 'Common.UsageType': string;
  readonly CurrencyCode: string;
  readonly 'Dynamic.TotalArea': string;
  readonly Interval: string;
  readonly PageSize: string;
  readonly RadiusType: string;
  readonly Site: string;
  readonly PropertyDefaultEnumerationSuffixes?: string[];
  readonly PropertyOrigins?: string[];
  readonly PropertyQualifier?: string;
  readonly Sort: string;
  readonly Unit: string;
}

export interface Polygon {
  coords: string,
  value: string,
  label: string,
  default?: boolean,
  name?: string
}

export interface SearchConfig {
  readonly searchResultsPage: string; // Can't be an empty string
  readonly searchHeaderLinkUrl?: string; // Can't be an empty string
  readonly searchBias?: SearchBias; // If it's defined biasRadius is compulsory
  readonly biasRadius?: number;
  readonly restrictToCountry?: string;
  readonly searchPlaceTypes?: string[]; // Can't be an empty array
  readonly allowAssumeZipCode?: boolean; // if this is present and true, then at 4-5 digits we add (regions) to types to eliminate IP location bias
  readonly searchPathSortAlphabetical?: boolean; // sort the search path selector alphabetically (i.e, property types)
  readonly searchHeader: SearchHeader;
  readonly hideSearchToBuy: boolean;
  readonly hideSearchToLet: boolean;
  readonly searchPathSelector?: SearchPath[]; // Can't be an empty array
  readonly propertyIdSearchPathSelector?: PropertyIdSearchPath[]; // Can't be an empty array
  readonly searchSuggestDebounce?: number;
  readonly polygons?: Polygon[];
}

export interface SearchPath {
  readonly value: string;
  readonly label: string;
  readonly default: boolean;
}

export interface PropertyIdSearchPath {
  readonly value: string;
  readonly label: string;
  readonly display?: string;
}

export interface SearchHeader {
  readonly searchHeaderTitleText?: string;
  readonly searchHeaderLinkText?: string;
  readonly searchHeaderLinkUrl?: string;
}

export interface SearchBias {
  readonly lat: string;
  readonly lon: string;
}

export interface LocationTypeDefinition {
  readonly name: string;
  readonly definitions: string[];
  readonly radius: number;
}

export interface OtherConfig {
  readonly name: string;
  readonly endpoint: string;
}

export enum TransactionTypeValue {
  isSale = 'isSale',
  isLetting = 'isLetting',
}

export enum DistanceUnit {
  miles = 'miles',
  km = 'km',
}
