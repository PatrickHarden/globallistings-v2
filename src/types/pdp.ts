export interface PdpPath {
  readonly propertyId: string;
}
