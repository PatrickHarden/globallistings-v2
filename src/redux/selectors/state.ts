import { State } from '../../types/state';
import { theme } from '../../themes/default/theme';
import { createSelector } from 'reselect';
import { FilterType } from '../../types/config';

// State
export const entityMetaSelector = (state: State) => state.entitiesMeta;
export const entitiesSelector = (state: State) => state.entities;
export const uiSelector = (state: State) => state.ui;
export const locationSelector = (state: State) => state.router.location;

// Entities
export const propertiesSelector = createSelector(
  entitiesSelector,
  entities => entities.properties
);

export const geocodeSelector = createSelector(
  entitiesSelector,
  entities => entities.geocode
);

export const propertySearchesSelector = createSelector(
  entitiesSelector,
  entities => entities.propertySearches
);

export const configSelector = createSelector(
  entitiesSelector,
  entities => entities.config
);

export const autocompleteSelector = createSelector(
  entitiesSelector,
  entities => entities.autocomplete
);

export const autocompleteSearchesSelector = createSelector(
  entitiesSelector,
  entities => entities.autocompleteSearches
);

// Meta
export const propertiesMetaSelector = createSelector(
  entityMetaSelector,
  meta => meta.propertySearches
);

export const geocodeMetaSelector = createSelector(
  entityMetaSelector,
  meta => meta.geocode
);

export const configMetaSelector = createSelector(
  entityMetaSelector,
  meta => meta.config
);

export const autocompleteSearchesMetaSelector = createSelector(
  entityMetaSelector,
  meta => meta.autocompleteSearches
);

// Config
export const localeSelector = createSelector(
  configSelector,
  config => config.language
);

export const messagesSelector = createSelector(
  configSelector,
  config => config.i18n
);

export const filtersSelector = createSelector(
  configSelector,
  config => config.filters
);

export const rangeFiltersSelector = createSelector(
  filtersSelector,
  filters => filters.filter(f => f.type === FilterType.range)
);

// Other

// TODO add logic to switch between themes based on config
export const themeSelector = (state: State) => theme;

// TODO figure out where it will be coming from
export const googleMapsApiKeySelector = () => 'AIzaSyCzEw2NaQFixR7sZCLnZ1oGQljW7AxLZ7k';
