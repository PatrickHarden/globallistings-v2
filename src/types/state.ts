import { RouterState } from 'connected-react-router';
import { Dictionary } from './common/dictionary';
import { Property } from './property';
import { Config, AppConfig } from './config';
import { UI } from './ui';
import { Autocomplete } from './autocomplete';

export interface State {
  readonly entities: Entities;
  readonly entitiesMeta: EntitiesMeta;
  readonly router: RouterState;
  readonly ui: UI;
}

export interface EntitiesMeta {
  readonly propertySearches: Dictionary<EntityMeta>;
  readonly geocode: Dictionary<EntityMeta>;
  readonly autocompleteSearches: Dictionary<EntityMeta>;
  readonly config: EntityMeta;
}

export type LoadStatus = 'PENDING' | 'ERROR' | 'SUCCESS' | 'NONE';

export interface EntityMeta {
  readonly status: LoadStatus;
}

export interface LatLng {
  readonly lat: number;
  readonly lng: number;
}

export interface Bounds {
  readonly sw: LatLng;
  readonly ne: LatLng;
}

export interface Geocode {
  readonly bounds: Bounds;
  readonly center: LatLng;
}

export interface Entities {
  readonly properties: Dictionary<Property>;
  readonly propertySearches: Dictionary<string[]>;
  readonly autocomplete: Dictionary<Autocomplete>;
  readonly autocompleteSearches: string[];
  readonly geocode: Dictionary<Geocode>;
  readonly config: AppConfig;
}

export function isConfigObject(config: string | Config): config is Config {
  return typeof config !== 'string';
}
