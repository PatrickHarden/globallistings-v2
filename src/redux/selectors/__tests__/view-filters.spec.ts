jest.mock('connected-react-router', () => {
  const original = require.requireActual('connected-react-router');
  return {
    ...original,
    createMatchSelector: () => () => ({
      params: { location: 'london', usageType: 'office', transactionType: 'sale' },
    }),
  };
});

import {
  primaryFiltersSelector,
  secondaryFiltersSelector,
  conditionalFiltersSelector,
} from '../view-filters';
import { DeepPartial } from 'redux';
import { State } from '../../../types/state';

describe(primaryFiltersSelector, () => {
  it('should retrieve all filters with a primary placement', () => {
    const state: DeepPartial<State> = {
      router: { location: { search: '?radius=1' } },
      entities: {
        config: {
          filters: [
            {
              name: 'filter1',
              type: 'select',
              placement: 'secondary',
              options: [
                { value: '', label: 'None' },
                { value: '1', label: 'One' },
                { value: '2', label: 'Two' },
              ],
            },
            {
              name: 'filter2',
              type: 'select',
              placement: 'primary',
              options: [
                { value: '', label: 'None' },
                { value: '1', label: 'One' },
                { value: '2', label: 'Two' },
              ],
            },
          ],
        },
      },
      ui: { search: {} },
    };

    const result = primaryFiltersSelector(state as State);

    expect(result.length).toBe(1);
    expect(result[0].name).toBe('filter2');
  });
});

describe(secondaryFiltersSelector, () => {
  it('should retrieve all filters with a secondary placement', () => {
    const state: DeepPartial<State> = {
      router: { location: { search: '?radius=1' } },
      entities: {
        config: {
          filters: [
            {
              name: 'filter1',
              type: 'select',
              placement: 'secondary',
              options: [
                { value: '', label: 'None' },
                { value: '1', label: 'One' },
                { value: '2', label: 'Two' },
              ],
            },
            {
              name: 'filter2',
              type: 'select',
              placement: 'primary',
              options: [
                { value: '', label: 'None' },
                { value: '1', label: 'One' },
                { value: '2', label: 'Two' },
              ],
            },
          ],
        },
      },
      ui: { search: {} },
    };

    const result = secondaryFiltersSelector(state as State);

    expect(result.length).toBe(1);
    expect(result[0].name).toBe('filter1');
  });
});

describe(conditionalFiltersSelector, () => {
  let state: DeepPartial<State>;
  beforeEach(() => {
    state = {
      router: {
        location: {
          pathname: '/usageType/sale/london/results',
          search: '',
          hash: '',
        },
        action: 'POP',
      },
      entities: {
        config: {
          params: {
            Site: 'foo',
            Sort: 'bar',
            Unit: 'baz',
          },
          filters: [
            {
              name: 'filter1',
              type: 'select',
              placement: 'secondary',
              options: [
                { value: '', label: 'None' },
                { value: '1', label: 'One' },
                { value: '2', label: 'Two' },
              ],
            },
            {
              name: 'filter2',
              type: 'select',
              placement: 'primary',
              options: [
                { value: '', label: 'None' },
                { value: '1', label: 'One' },
                { value: '2', label: 'Two' },
              ],
            },
          ],
        },
        geocode: {
          london: {
            bounds: {
              ne: { lat: 1, lng: 2 },
              sw: { lat: 1, lng: 2 },
            },
            center: {
              lat: 1,
              lng: 2,
            },
          },
        },
      },
      ui: { search: {} },
    };
  });

  it('should return all flters when none have conditions attached', () => {
    const result = conditionalFiltersSelector(state as State);
    expect(result.length).toBe(2);
  });

  it('should return all flters with no conditions attached and filters that meet conditions', () => {
    state.entities.config.filters.push({
      name: 'filter3',
      type: 'select',
      placement: 'primary',
      conditional: {
        Sort: 'bar',
      },
      options: [
        { value: '', label: 'None' },
        { value: '1', label: 'One' },
        { value: '2', label: 'Two' },
      ],
    });
    const result = conditionalFiltersSelector(state as State);
    expect(result.length).toBe(3);
  });

  it('should return all flters with no conditions attached and filters that meet conditions but remove filters that dont meet conditions', () => {
    state.entities.config.filters.push(
      {
        name: 'filter3',
        type: 'select',
        placement: 'primary',
        conditional: {
          Sort: 'foo',
        },
        options: [
          { value: '', label: 'None' },
          { value: '1', label: 'One' },
          { value: '2', label: 'Two' },
        ],
      },
      {
        name: 'filter4',
        type: 'select',
        placement: 'primary',
        conditional: {
          Sort: 'bar',
        },
        options: [
          { value: '', label: 'None' },
          { value: '1', label: 'One' },
          { value: '2', label: 'Two' },
        ],
      }
    );
    const result = conditionalFiltersSelector(state as State);
    expect(result.length).toBe(3);
  });

  it('will return a filter if ALL conditions pass', () => {
    state.entities.config.filters.push({
      name: 'filter3',
      type: 'select',
      placement: 'primary',
      conditional: {
        Site: 'foo',
        Sort: 'bar',
      },
      options: [
        { value: '', label: 'None' },
        { value: '1', label: 'One' },
        { value: '2', label: 'Two' },
      ],
    });
    const result = conditionalFiltersSelector(state as State);
    expect(result.length).toBe(3);
  });

  it('will NOT return a filter if not all conditions pass', () => {
    state.entities.config.filters.push({
      name: 'filter3',
      type: 'select',
      placement: 'primary',
      conditional: {
        Site: 'foo',
        Sort: 'baz',
      },
      options: [
        { value: '', label: 'None' },
        { value: '1', label: 'One' },
        { value: '2', label: 'Two' },
      ],
    });
    const result = conditionalFiltersSelector(state as State);
    expect(result.length).toBe(2);
  });
});

jest.mock('connected-react-router', () => {
  const original = require.requireActual('connected-react-router');
  return {
    ...original,
    createMatchSelector: () => () => ({
      params: { location: 'london', usageType: 'industrial', transactionType: 'letting' },
    }),
  };
});

describe(conditionalFiltersSelector, () => {
  let state: DeepPartial<State>;

  it('make sure selected items are returned to properly set filters', () => {
    const state: DeepPartial<State> = {
      router: {
        location: {
          pathname: '/industrial/letting/london/results',
          search: '',
          hash: '',
        },
        action: 'POP',
      },
      entities: {
        config: {
          filters: [
            {
              name: 'usageType',
              type: 'select',
              options: [
                { value: 'office', label: 'Office' },
                { value: 'industrial', label: 'Industrial' },
                { value: 'retail', label: 'Retail' },
              ],
            },
            {
              name: 'transactionType',
              type: 'select',
              options: [{ value: 'sale', label: 'To Buy' }, { value: 'letting', label: 'To Let' }],
            },
          ],
        },
        geocode: {
          london: {
            bounds: {
              ne: { lat: 1, lng: 2 },
              sw: { lat: 1, lng: 2 },
            },
            center: {
              lat: 1,
              lng: 2,
            },
          },
        },
      },
      ui: { search: {} },
    };
    expect(conditionalFiltersSelector(state as State)).toMatchSnapshot();
  });
});
