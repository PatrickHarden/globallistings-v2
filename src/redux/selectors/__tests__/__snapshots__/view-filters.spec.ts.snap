// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`make sure selected items are returned to properly set filters 1`] = `
Array [
  Object {
    "items": Array [
      Object {
        "label": "Office",
        "value": "office",
      },
      Object {
        "label": "Industrial",
        "value": "industrial",
      },
      Object {
        "label": "Retail",
        "value": "retail",
      },
    ],
    "name": "usageType",
    "options": Array [
      Object {
        "label": "Office",
        "value": "office",
      },
      Object {
        "label": "Industrial",
        "value": "industrial",
      },
      Object {
        "label": "Retail",
        "value": "retail",
      },
    ],
    "selectedItem": Object {
      "label": "Industrial",
      "value": "industrial",
    },
    "type": "select",
  },
  Object {
    "items": Array [
      Object {
        "label": "To Buy",
        "value": "sale",
      },
      Object {
        "label": "To Let",
        "value": "letting",
      },
    ],
    "name": "transactionType",
    "options": Array [
      Object {
        "label": "To Buy",
        "value": "sale",
      },
      Object {
        "label": "To Let",
        "value": "letting",
      },
    ],
    "selectedItem": Object {
      "label": "To Let",
      "value": "letting",
    },
    "type": "select",
  },
]
`;
